File: typescript/src/core/engines/pruner.ts
// Engine B: IDF-based token pruning — remove low-information words.

import type { MemoSiftConfig } from "../../config.js";
import {
  type AnchorLedger,
  type ClassifiedMessage,
  CompressionPolicy,
  createClassified,
  createMessage,
} from "../types.js";

const TARGET_POLICIES = new Set([CompressionPolicy.MODERATE, CompressionPolicy.AGGRESSIVE]);

const PROTECTED_PATTERNS = [
  /^[\w.\-]+(?:[/\\][\w.\-]+)+(?::\d+)?$/, // file paths
  /^\d[\d.,:]*$/, // numbers
  /^[a-z]+(?:[A-Z][a-z]+)+$/, // camelCase
  /^[a-z]+(?:_[a-z]+)+$/, // snake_case
  /^[A-Z][A-Z_]+$/, // UPPER_CASE
  /^https?:\/\/\S+$/, // URLs
  /^\S+@\S+\.\S+$/, // emails
  /^[A-Za-z0-9]{10,}$/, // tracking numbers
  /^(?:ORD|RET|INV|TXN|REF)-[\w\-]+$/, // structured IDs
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i, // UUIDs
  /^[A-Z]{2,4}-\d{4,}$/, // Order IDs
  /^[A-Za-z0-9]{12,}$/, // Long alphanumeric identifiers
];

export function pruneTokens(
  segments: ClassifiedMessage[],
  config: MemoSiftConfig,
  ledger?: AnchorLedger | null,
): ClassifiedMessage[] {
  const allContents = segments.map((s) => s.message.content);
  const idfScores = computeIdfScores(allContents);

  // Auto-protect tokens from anchor ledger (Item 1.3).
  const ledgerLower = new Set<string>();
  if (ledger) {
    for (const s of ledger.getProtectedStrings()) {
      ledgerLower.add(s.toLowerCase());
    }
  }

  return segments.map((seg) => {
    if (!TARGET_POLICIES.has(seg.policy)) return seg;
    const pruned = pruneSegment(
      seg.message.content,
      idfScores,
      config.tokenPruneKeepRatio,
      ledgerLower,
    );
    if (pruned === seg.message.content) return seg;
    const newMsg = createMessage(seg.message.role, pruned, {
      name: seg.message.name,
      toolCallId: seg.message.toolCallId,
      toolCalls: seg.message.toolCalls,
      metadata: seg.message.metadata,
    });
    return createClassified(newMsg, seg.contentType, seg.policy, {
      originalIndex: seg.originalIndex,
      relevanceScore: seg.relevanceScore,
      protected: seg.protected,
      importanceScore: seg.importanceScore,
      shield: seg.shield,
    });
  });
}

function computeIdfScores(documents: string[]): Map<string, number> {
  const nDocs = documents.length;
  const docFreq = new Map<string, number>();
  for (const doc of documents) {
    const unique = new Set(doc.toLowerCase().match(/\b\w+\b/g) ?? []);
    for (const token of unique) docFreq.set(token, (docFreq.get(token) ?? 0) + 1);
  }
  const idf = new Map<string, number>();
  for (const [token, freq] of docFreq) {
    idf.set(token, Math.log((nDocs + 1) / (freq + 1)) + 1);
  }
  return idf;
}

function isProtectedToken(token: string, ledgerLower: ReadonlySet<string>): boolean {
  if (PROTECTED_PATTERNS.some((p) => p.test(token))) return true;
  if (ledgerLower.size > 0 && ledgerLower.has(token.toLowerCase())) return true;
  return false;
}

function pruneSegment(
  text: string,
  idfScores: Map<string, number>,
  keepRatio: number,
  ledgerLower: ReadonlySet<string>,
): string {
  return text
    .split("\n")
    .map((line) => {
      const stripped = line.trim();
      if (!stripped) return line;
      const words = stripped.split(/\s+/);
      if (words.length === 0) return line;

      const scored = words.map((w) => ({
        word: w,
        score: idfScores.get(w.toLowerCase().replace(/\W/g, "")) ?? 1.0,
      }));
      const sorted = [...scored].sort((a, b) => a.score - b.score);
      const keepCount = Math.max(1, Math.floor(sorted.length * keepRatio));
      if (keepCount >= sorted.length) return line;
      const threshold = sorted[sorted.length - keepCount]!.score;

      const kept = scored
        .filter((s) => isProtectedToken(s.word, ledgerLower) || s.score >= threshold)
        .map((s) => s.word);
      return kept.length > 0 ? kept.join(" ") : line;
    })
    .join("\n");
}

File: python/src/memosift/core/engines/pruner.py
# Engine B: IDF-based token pruning — remove low-information words.
from __future__ import annotations

import re
from dataclasses import replace as dc_replace
from math import log
from typing import TYPE_CHECKING

from memosift.core.types import (
    AnchorLedger,
    ClassifiedMessage,
    CompressionPolicy,
    MemoSiftMessage,
)

if TYPE_CHECKING:
    from memosift.config import MemoSiftConfig

# Applies to TOOL_RESULT_TEXT, ASSISTANT_REASONING.
_TARGET_POLICIES = {
    CompressionPolicy.MODERATE,
    CompressionPolicy.AGGRESSIVE,
}

# Protected token patterns — never pruned regardless of IDF score.
_PROTECTED_PATTERNS: list[re.Pattern[str]] = [
    re.compile(r"(?:[A-Za-z]:)?[\w.\-]+(?:[/\\][\w.\-]+)+(?::\d+)?"),  # File paths
    re.compile(r"\b\d[\d.,:]*\b"),  # Numbers, line numbers, error codes
    re.compile(r"\b[a-z]+(?:[A-Z][a-z]+)+\b"),  # camelCase
    re.compile(r"\b[a-z]+(?:_[a-z]+)+\b"),  # snake_case
    re.compile(r"\b[A-Z][A-Z_]+\b"),  # UPPER_CASE
    re.compile(r"https?://\S+"),  # URLs
    re.compile(r"\S+@\S+\.\S+"),  # Email addresses
    re.compile(r"\b[A-Za-z0-9]{10,}\b"),  # Tracking numbers (e.g., 1Z999AA10123456784)
    re.compile(r"\b(?:ORD|RET|INV|TXN|REF)-[\w\-]+\b"),  # Structured IDs (ORD-2026-78432)
    re.compile(
        r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}", re.IGNORECASE
    ),  # UUIDs
    re.compile(r"\b[A-Z]{2,4}-\d{4,}\b"),  # Order IDs (AB-12345)
    re.compile(r"\b[A-Za-z0-9]{12,}\b"),  # Alphanumeric identifiers ≥12 chars
]


def prune_tokens(
    segments: list[ClassifiedMessage],
    config: MemoSiftConfig,
    ledger: AnchorLedger | None = None,
) -> list[ClassifiedMessage]:
    """Prune low-information tokens from eligible segments using IDF scoring.

    High-IDF tokens (appearing in few messages) are information-dense and kept.
    Low-IDF tokens (appearing in many messages) are noise and pruned.
    Protected tokens (file paths, identifiers, numbers, anchor ledger facts)
    are never pruned.

    Args:
        segments: Classified messages from previous layers.
        config: Pipeline configuration (controls ``token_prune_keep_ratio``).
        ledger: Optional anchor ledger — tokens matching ledger facts are protected.

    Returns:
        Segments with low-information tokens removed.
    """
    # Compute IDF scores across all message contents.
    all_contents = [seg.content for seg in segments]
    idf_scores = _compute_idf_scores(all_contents)
    # Auto-protect tokens that appear in the anchor ledger (Item 1.3).
    # This ensures file paths, error messages, IDs, and other critical tokens
    # extracted into the ledger are never pruned from the source text.
    ledger_lower: frozenset[str] = frozenset()
    if ledger is not None:
        ledger_lower = frozenset(s.lower() for s in ledger.get_protected_strings())

    result: list[ClassifiedMessage] = []
    for seg in segments:
        if seg.policy in _TARGET_POLICIES:
            pruned = _prune_segment(
                seg.content, idf_scores, config.token_prune_keep_ratio, ledger_lower
            )
            if pruned != seg.content:
                new_msg = MemoSiftMessage(
                    role=seg.message.role,
                    content=pruned,
                    name=seg.message.name,
                    tool_call_id=seg.message.tool_call_id,
                    tool_calls=seg.message.tool_calls,
                    metadata=seg.message.metadata,
                )
                result.append(dc_replace(seg, message=new_msg))
            else:
                result.append(seg)
        else:
            result.append(seg)
    return result


def _compute_idf_scores(documents: list[str]) -> dict[str, float]:
    """Compute IDF scores across the message corpus.

    IDF = log((N + 1) / (df + 1)) + 1
    """
    n_docs = len(documents)
    doc_freq: dict[str, int] = {}
    for doc in documents:
        unique_tokens = set(re.findall(r"\b\w+\b", doc.lower()))
        for token in unique_tokens:
            doc_freq[token] = doc_freq.get(token, 0) + 1
    return {token: log((n_docs + 1) / (freq + 1)) + 1 for token, freq in doc_freq.items()}


def _is_protected_token(
    token: str,
    ledger_lower: frozenset[str] = frozenset(),
) -> bool:
    """Return True if token matches any protected pattern or anchor ledger fact."""
    for pattern in _PROTECTED_PATTERNS:
        if pattern.fullmatch(token):
            return True
    # Check if the token matches any ledger critical string (pre-lowered).
    return bool(ledger_lower and token.lower() in ledger_lower)


def _prune_segment(
    text: str,
    idf_scores: dict[str, float],
    keep_ratio: float,
    ledger_lower: frozenset[str] = frozenset(),
) -> str:
    """Prune low-IDF tokens from a text segment.

    Processes line-by-line to preserve structure.
    Protected tokens and tokens with above-threshold IDF are kept.
    """
    lines = text.split("\n")
    result_lines: list[str] = []

    for line in lines:
        stripped = line.strip()
        if not stripped:
            result_lines.append(line)
            continue

        # Tokenize preserving whitespace structure.
        tokens = re.findall(r"\S+|\s+", line)
        word_tokens = [(t, idf_scores.get(t.lower().strip(), 1.0)) for t in tokens if t.strip()]

        if not word_tokens:
            result_lines.append(line)
            continue

        # Determine keep threshold based on keep_ratio.
        scores = sorted([s for _, s in word_tokens])
        keep_count = max(1, int(len(scores) * keep_ratio))
        if keep_count >= len(scores):
            result_lines.append(line)
            continue

        threshold = scores[len(scores) - keep_count]

        # Keep tokens above threshold or that are protected.
        kept: list[str] = []
        for token, score in word_tokens:
            if _is_protected_token(token, ledger_lower) or score >= threshold:
                kept.append(token)

        if kept:
            result_lines.append(" ".join(kept))
        # Skip entirely empty lines from pruning.

    return "\n".join(result_lines)
